students/JeroenSweerts/Lesson6/mailroom4.py

```diff
@@ -0,0 +1,105 @@
+    try:
+        donation = input('Please enter a donation amount > ')
+        donations[name].append(int(donation))
+    except ValueError:
+        while not donation.isdigit():
+            print("You didn't enter a valid amount.")
+            donation = input('Please enter a donation amount > ')
+        donations[name].append(int(donation))
+    d = {'name':name, 'donation':donation}
+    print("Dear {name} , thank you very much for your generous donation of {donation}. Looking forward receiving even more money next time.".format(**d))
+    print(donations.keys())
+    print(donations.values())
+
+
+def createfile(name):
+    '''create .txt file containing thank you letter for each person who made a donation'''
+    d = {'name':name, 'donation':donations[name][-1]}
+    outfile = open(name+'.txt', 'w')
+    outfile.write("Dear {name} , \n\n".format(**d))
+    outfile.write("\t Thank you very much for your generous donation of ${donation}. \n\n".format(**d))
+    outfile.write("\t It will be put to very good use. \n\n")
+    outfile.write("\t\t Sincerely, \n")
+    outfile.write("\t\t   -The Team \n")
+    outfile.close()
+
+
+def thankyoueveryone():
+    [createfile(name) for name in donations.keys()]
+
+
+def report():
+    """ If the user (you) selected “Create a Report”, print a list of your donors,
+     sorted by total historical donation amount.
+     Include Donor Name, total donated, number of donations and average donation
+     amount as values in each row. You do not need to print out all their
+     donations, just the summary info.
+     Using string formatting, format the output rows as nicely as possible.
+     The end result should be tabular (values in each column should align with
+     those above and below)
+     After printing this report, return to the original prompt.
+     At any point, the user should be able to quit their current task and return
+     to the original prompt.
+     From the original prompt, the user should be able to quit the script cleanly."""
+    print(f'{"Donor Name":20s} {"|  Total Given":20s} {"|  Num Gifts  |":20s} {"Average Gift":20s}')
+    print(f'{"-"*76}')
+    [print(f'{name:20s} ${sum(donations[name]):20.2f} {len(donations[name]):13d}${sum(donations[name])/len(donations[name]):20.2f}') for name in donations.keys()]
+
+
+if __name__ == '__main__':
+    response = input('Please choose between the following 3 actions:\
+    1. Send a Thank You; 2. Create a Report; 3. Send letters to everyone; 4. quit >  ')
+    try:
+        int(response)
+    except:
+        while not response.isdigit():
+            print("You didn't enter a number.")
+            print()
+            response = input('Please choose between the following 3 actions:\
+            1. Send a Thank You; 2. Create a Report; 3. Send letters to everyone; 4. quit >  ')
+    arg_dict = {1:thankyou, 2:report, 3:thankyoueveryone}
+    while int(response) != 4:
+        try:
+            arg_dict.get(int(response))()
```

Comment:
I think the only error that you need to catch here is `KeyError`, and that would let you eliminate a lot of the added validation and repeated user prompts because this will be raised by any input value that is not in the `arg_dict`